wal/encoder.go

```diff
@@ -45,9 +48,23 @@ func (e *encoder) encode(rec *walpb.Record) error {
 
 	e.crc.Write(rec.Data)
 	rec.Crc = e.crc.Sum32()
-	data, err := rec.Marshal()
-	if err != nil {
-		return err
+	var (
+		data []byte
+		err  error
+		n    int
+	)
+
+	if rec.Size() > len(e.buf) {
+		data, err = rec.Marshal()
+		if err != nil {
+			return err
+		}
+	} else {
+		n, err = rec.MarshalTo(e.buf)
+		if err != nil {
+			return err
+		}
+		data = e.buf[:n]
```

Comment:
What about making this a function called something like `buf = MaybeMarshal(obj, best_effort_buf)`?
I think this pattern will be really common in the future optimization.